File: src/main/java/dev/akif/crud/CRUDError.java
package dev.akif.crud;

import jakarta.validation.constraints.NotNull;
import org.springframework.http.HttpStatus;

/**
 * Error model of the API operations
 *
 * @param code    Numeric code matching the HTTP status code of the response this error should produce
 * @param message Error message describing what went wrong
 * @param type    Error type matching the HTTP status of the response this error should produce
 */
public record CRUDError(
        @NotNull int code,
        @NotNull String message,
        @NotNull String type
) {
    /**
     * Constructor to create a {@link CRUDError} with given data
     *
     * @param status  {@link HttpStatus} to use for the response this error should produce
     * @param message Error message describing what went wrong
     */
    public CRUDError(final HttpStatus status, final String message) {
        this(status.value(), message, status.getReasonPhrase());
    }
}

File: api/src/main/kotlin/dev/akif/crud/error/CRUDError.kt
package dev.akif.crud.error

import com.fasterxml.jackson.annotation.JsonIgnore
import org.springframework.http.HttpStatus

/**
 * Error model of the API operations
 *
 * @property code    Numeric code matching the HTTP status code of the response this error should produce
 * @property message Error message describing what went wrong
 * @property type    Error type matching the HTTP status of the response this error should produce
 */
@JvmRecord
data class CRUDError(
    @JsonIgnore @JvmField val code: Int,
    @JvmField val message: String,
    @JvmField val type: String
) {
    /**
     * Constructor to create a [CRUDError] with given data
     *
     * @param status  [HttpStatus] to use for the response this error should produce
     * @param message Error message describing what went wrong
     */
    constructor(status: HttpStatus, message: String) : this(status.value(), message, status.reasonPhrase)
}
